docs/proposal/20220608-capa-ami-github-action.md

```diff
@@ -0,0 +1,346 @@
+---
+title: CAPA GitHub Action for Building/Publishing AMIs
+authors:
+  - "@zeborg"
+reviewers:
+  - "@sedefsavas"
+  - "@richardcase"
+creation-date: 2022-06-08
+last-updated: 2022-06-08
+status: implementable
+see-also:
+- https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1982
+replaces: []
+superseded-by: []
+---
+
+# CAPA GitHub Action for Building/Publishing AMIs
+
+## Table of Contents
+
+- [CAPA GitHub Action for Building/Publishing AMIs](#capa-github-action-for-buildingpublishing-amis)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+  - [Proposal](#proposal)
+    - [Flowchart](#flowchart)
+    - [Components](#components)
+  - [Testing the CAPA GitHub Action](#testing-the-capa-github-action)
+  - [Migrating the Presubmit and Postsubmit Workflows to Prow](#migrating-the-presubmit-and-postsubmit-workflows-to-prow)
+  - [Roadblocks & Limitations](#roadblocks--limitations)
+  - [Scope of Improvement](#scope-of-improvement)
+
+## Summary
+The Cluster API Provider AWS (CAPA) project builds and publishes AMIs at every new Kubernetes release for the 3 most recent minor versions of Kubernetes. This proposal is aimed towards making this entire process automated over cloud instead of a user’s system by using GitHub Actions.
+
+Tested repo: https://github.com/zeborg/cluster-api-provider-aws/tree/ami-action
+
+## Motivation
+At present, this process of building and publishing the AMIs is carried out manually by one of the core project members on their own computer system. The usual steps to perform this task are:
+1. Setup [image-builder](https://github.com/kubernetes-sigs/image-builder/) and its dependencies on their system.
+2. Setup AWS CLI and the AWS account’s credentials on which the images are supposed to be published.
+3. Configure the `image-builder` packer flags to customize the image specifications (eg. Kubernetes version, CNI version, etc.) using a packer configuration file that stores these custom specifications as variables within a JSON file as key-value pairs.
+4. Run `image-builder`, which may take up to 5 hours to build the images for all the supported OS and copy them to all the supported regions.
```

Comment:
5 hours???!! 😨

Reply:
Might be building serially for each OS type and each latest k8s patch release.
@zeborg Could we see if this can be parallelized atleast for different k8s versions if not for OS types?

Reply:
@Ankitasw - tbh I didn't consider that yet, but it sounds like a really good idea! I'll start working on updating the existing design and workflow for this.